### PR TITLE
Add support for Barefoot Tofino model

### PR DESCRIPTION
**Pull request by bocon13**
Tested against SDE 9.3.0

### CODE REVIEW

File: stratum/hal/bin/barefoot/docker/Dockerfile.model [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    procps \
+    libcli1.9
+RUN apt-get update && \
+    apt-get install -y $BUILD_DEPS && \
+    rm -rf /var/lib/apt/lists/*
+
+ARG SDE_TAR_NAME
+ARG SDE_VER
+
+RUN ls
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/Dockerfile.model:20]: remove

File: stratum/hal/bin/barefoot/docker/Dockerfile.model [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ADD ./${SDE_TAR_NAME} /
+ADD ./tofino_skip_p4.conf /usr/share/stratum/
+ADD ./start-tofino-model.sh /usr/local/bin/
+
+RUN tar -xzvf bf-sde-$SDE_VER/packages/tofino-model-$SDE_VER.tgz \
+    && tar -xzvf bf-sde-$SDE_VER/packages/ptf-modules-$SDE_VER.tgz \
+    && rm -rf bf-sde-$SDE_VER \
+    && cp tofino-model-$SDE_VER/tofino-model.x86_64.bin /usr/local/bin/tofino-model \
+    && cp ptf-modules-$SDE_VER/ptf-utils/veth_setup.sh /usr/local/bin/ \
+    && cp ptf-modules-$SDE_VER/ptf-utils/dma_setup.sh /usr/local/bin/ \
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/Dockerfile.model:30]: should we move this into start-stratum?

File: stratum/hal/bin/barefoot/docker/Dockerfile.model [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+RUN apt-get update && \
+    apt-get install -y $BUILD_DEPS && \
+    rm -rf /var/lib/apt/lists/*
+
+ARG SDE_TAR_NAME
+ARG SDE_VER
+
+RUN ls
+ADD ./${SDE_TAR_NAME} /
+ADD ./tofino_skip_p4.conf /usr/share/stratum/
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/Dockerfile.model:22]: Add the no_bsp.conf as well?

File: stratum/hal/bin/barefoot/docker/build-tofino-model-container.sh [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  print_help
+  exit 1
+fi
+
+DOCKER_EXTRA_RUN_OPTS=""
+if [ -t 0 ]; then
+  # Running in a TTY, so run interactively (i.e. make Ctrl-C work)
+  DOCKER_EXTRA_RUN_OPTS+="-it "
+fi
+
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/build-tofino-model-container.sh:30]:
```suggestion
```

File: stratum/hal/bin/barefoot/start-tofino-model.sh [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright 2018-present Open Networking Foundation
+# SPDX-License-Identifier: Apache-2.0
+set -e
+
+veth_setup.sh ||
+( echo "Failed to set up interfaces. You need to run in priveledged mode.";
+  exit 1 )
+
+dma_setup.sh
+tofino-model --p4-target-config /usr/share/stratum/tofino_skip_p4.conf $@
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/start-tofino-model.sh:11]:
```suggestion
exec tofino-model --p4-target-config /usr/share/stratum/tofino_skip_p4.conf $@
```

File: stratum/hal/bin/barefoot/start-tofino-model.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# Copyright 2018-present Open Networking Foundation
+# SPDX-License-Identifier: Apache-2.0
+set -e
+
+veth_setup.sh ||
+( echo "Failed to set up interfaces. You need to run in priveledged mode.";
+  exit 1 )
+
+dma_setup.sh
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/start-tofino-model.sh:10]: consider moving this into start-stratum

File: stratum/hal/bin/barefoot/README.run.md [hunk 7]
```diff
@@ -278,6 +278,27 @@ The `--bf_sim` flag tells Stratum not to use the Phal ONLP implementation, but
 `PhalSim`, a "fake" Phal implementation, instead. Use this flag when you are
 using a vendor-provided BSP or running Stratum with the Tofino software model.
 
+
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/README.run.md:281]: TODO add more context for the readme

File: stratum/hal/bin/barefoot/README.run.md [hunk 8]
```diff
@@ -278,6 +278,27 @@ The `--bf_sim` flag tells Stratum not to use the Phal ONLP implementation, but
...
+```bash
+docker run --rm -it --privileged \
+  --network=host \
+  stratumproject/tofino-model:9.3.0
+```
+
+```bash
+docker run -it --rm --privileged \
+  --env PLATFORM=barefoot-tofino-model \
+  --network=host \
```
Comment (ccascone) [inline stratum/hal/bin/barefoot/README.run.md:295]: The downside of using `host` is a higher chance of port conflicts. The alternative is using `--network "container:tofino-model"`

**Comment by bocon13** [comment]
Need to mount the p4 compiler output for logging and refer to the right paths in conf file.

File: stratum/hal/bin/barefoot/docker/Dockerfile.model [hunk 9]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2012-present Open Networking Foundation
+# SPDX-License-Identifier: Apache-2.0
+
+FROM ubuntu:18.04
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/Dockerfile.model:4]: Can we find a smaller image/

File: stratum/hal/bin/barefoot/deb/start-stratum.sh [hunk 10]
```diff
@@ -33,9 +33,10 @@ fi
 PORT_MAP="/etc/stratum/$PLATFORM/port_map.json"
 if [ ! -f "$PORT_MAP" ]; then
     echo "Cannot find port map file $PORT_MAP"
-    exit 255
+    #exit 255
+else
+    ln -f -s "$PORT_MAP" /usr/share/port_map.json
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/deb/start-stratum.sh:38]: Can we ship a port_map under the fake `barefoot-tofino-model` platform?

File: stratum/hal/config/config_validator_main.cc [hunk 11]
```diff
@@ -27,12 +27,14 @@ TEST(ConfigValidator, TestChassisConfig) {
   EXPECT_OK(ReadProtoFromTextFile(filename, &chassis_config));
 }
 
+#ifndef SIM_TARGET
 TEST(ConfigValidator, TestPhalConfig) {
   std::string filename = absl::StrFormat(
       "stratum/hal/config/%s/phal_config.pb.txt", STRINGIFY(PLATFORM));
   PhalInitConfig phal_config;
   EXPECT_OK(ReadProtoFromTextFile(filename, &phal_config));
 }
```
Comment (pudelkoM) [inline stratum/hal/config/config_validator_main.cc:36]: This test should be optional anyway. A phal init config is only needed when the platform structure is not discoverable (i.e. broken or no ONLP).

File: stratum/hal/bin/barefoot/docker/Dockerfile.model (hunk 2 above)
Comment (pudelkoM) [reply stratum/hal/bin/barefoot/docker/Dockerfile.model:30]: The copy part or the run part?

**Comment by pudelkoM** [comment]
> Base automatically changed from sde-9.3.0 to master 4 hours ago

Nice

File: stratum/hal/bin/barefoot/start-tofino-model.sh (hunk 5 above)
Comment (bocon13) [reply stratum/hal/bin/barefoot/start-tofino-model.sh:11]: Don't do this. Ctrl-C isn't handled properly

File: stratum/hal/bin/barefoot/README.run.md (hunk 8 above)
Comment (bocon13) [reply stratum/hal/bin/barefoot/README.run.md:296]: Yes, this makes sense.

File: stratum/hal/bin/barefoot/README.run.md [hunk 12]
```diff
@@ -174,6 +174,65 @@ journalctl -u stratum_bf.service
...
+PLATFORM=barefoot-tofino-model \
+stratum/hal/bin/barefoot/docker/start-stratum-container.sh \
+  -bf_sim \
+  -bf_switchd_background=false \
+  -enable_onlp=false
+```
+
+### Cleaning up `tofino-model` interfaces
+
+To remove the interfaces created by when `tofino-model` starts, you can run
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/README.run.md:217]:
```suggestion
To remove the interfaces created when `tofino-model` starts, you can run
```

File: stratum/hal/bin/barefoot/README.run.md [hunk 13]
```diff
@@ -174,6 +174,65 @@ journalctl -u stratum_bf.service
...
+
+### Other deployment options
+
+You can run both Stratum and `tofino-model` natively on the host. They
+communicate over localhost TCP ports.
+
+If you wish to run multiple instances of Stratum and `tofino-model` on the
+same machine, you can use Docker's container network to link the containers
+together. For example, you can pass `--network container:stratum` when starting
+the `tofino-model` container.
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/README.run.md:233]:
```suggestion
same machine, you can use Docker's container network to link the containers'
networking stacks together. For example, you can pass `--network container:stratum`
when starting the `tofino-model` container.
```

File: stratum/hal/bin/barefoot/deb/start-stratum.sh (hunk 10 above)
Comment (bocon13) [reply stratum/hal/bin/barefoot/deb/start-stratum.sh:63]: I don't think a port map makes much sense for tofino-model. It's supposed to be in lieu of the BSP.

File: stratum/hal/bin/barefoot/docker/start-stratum-container.sh [hunk 14]
```diff
@@ -11,6 +16,21 @@ elif [[ -z "$PLATFORM" ]]; then
     exit 255
 fi
 
+if [[ "$PLATFORM" == 'barefoot-tofino-model' ]]; then
+    DOCKER_NET_OPTS="--network host"
+#    container_ids=$(docker ps | grep stratumproject/tofino-model | cut -d" " -f1)
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/start-stratum-container.sh:21]: clean this up

**Comment by bocon13** [comment]
addresses #370

File: stratum/hal/bin/barefoot/deb/start-stratum.sh [hunk 15]
```diff
@@ -79,7 +80,7 @@ if [ -f "$KDRV_PATH" ]; then
         exit 255
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/deb/start-stratum.sh:80]: ^^ this check is always true

File: stratum/hal/bin/barefoot/docker/start-tofino-model.sh [hunk 16]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+# Copyright 2018-present Open Networking Foundation
+# SPDX-License-Identifier: Apache-2.0
+set -e
+
+veth_setup.sh ||
+( echo "Failed to set up interfaces. You need to run in priveledged mode.";
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/start-tofino-model.sh:7]: check this

File: stratum/hal/bin/barefoot/docker/Dockerfile.model (hunk 3 above)
Comment (bocon13) [reply stratum/hal/bin/barefoot/docker/Dockerfile.model:32]: not needed... no port map for tofino model

File: stratum/hal/bin/barefoot/docker/start-stratum-container.sh [hunk 17]
```diff
@@ -11,6 +16,15 @@ elif [[ -z "$PLATFORM" ]]; then
     exit 255
 fi
 
+# Set Docker network options.
+if [[ "$PLATFORM" == 'barefoot-tofino-model' ]]; then
+    DOCKER_NET_OPTS="--network host"
```
Comment (bocon13) [inline stratum/hal/bin/barefoot/docker/start-stratum-container.sh:21]:
```suggestion
    DOCKER_NET_OPTS="--network host "
```

File: stratum/hal/bin/barefoot/docker/build-tofino-model-container.sh [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+MODEL_IMAGE=stratumproject/tofino-model:$SDE_VERSION
+echo "Building Tofino model runtime image: $RUNTIME_IMAGE"
+set -x
+docker build \
+  $DOCKER_BUILD_OPTS \
+  -t "$MODEL_IMAGE" \
+  --build-arg SDE_TAR_NAME="$SDE_TAR_NAME" \
+  --build-arg SDE_VER="$SDE_VERSION" \
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/docker/build-tofino-model-container.sh:47]: Add metadata labels?

Comment (bocon13) [reply stratum/hal/bin/barefoot/docker/build-tofino-model-container.sh:47]: Updated
```
            "Labels": {
                "bf-sde-version": "9.3.0",
                "build-machine": "menlo-pdp-lotta-nics",
                "build-timestamp": "2020-12-17T20:27:48+0000",
                "description": "This Docker image includes the Barefoot Tofino model",
                "maintainer": "Stratum dev <stratum-dev@lists.stratumproject.org>"
            }
```

**Comment by stolsma** [comment]
@bocon13 @pudelkoM Nice! Thanks for the continuous development this year!!! 🎉👍

**Pull request merged**